src/vmm/src/device_manager/resources.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cell::RefCell;
+
+pub use vm_allocator::AllocPolicy;
+use vm_allocator::{AddressAllocator, IdAllocator};
+
+use crate::arch;
+
+/// A resource manager for (de)allocating interrupt lines (GSIs) and guest memory
+///
+/// At the moment, we support:
+///
+/// * GSIs for legacy x86_64 devices
+/// * GSIs for MMIO devicecs
+/// * Memory allocations in the MMIO address space
+#[derive(Debug)]
+pub struct ResourceAllocator {
+    // Allocator for device interrupt lines
+    gsi_allocator: RefCell<IdAllocator>,
+    // Allocator for memory in the MMIO address space
+    mmio_memory: RefCell<AddressAllocator>,
+    // Memory allocator for ACPI data
+    #[cfg(target_arch = "x86_64")]
+    acpi_memory: RefCell<AddressAllocator>,
+}
```

Comment:
I kind of agree with you that it's not ideal to have the `Rc` separated than the `RefCell`. The main reason why I went with this design was that I wanted to avoid to have to deal with lifetimes. `MMIODeviceManager` and `ACPIDeviceManager` (in the next PR) get to store a reference to the allocator. I am willing to re-consider this, although then I'd probably prefer to break down this change in a separate PR before sending the ACPI PR in.

Speaking about "unidiomatism", do you see a practical problem with the way things are? The one I see is that, in theory, someone could create a `ResourceAllocator` without putting it inside an `Rc` and I don't know how would that play out. But since this is not a library that arbitrary consumers might use `ResourceAllocator`, I considered it ok.

Reply:
> The main reason why I went with this design was that I wanted to avoid to have to deal with lifetimes. MMIODeviceManager and ACPIDeviceManager (in the next PR) get to store a reference to the allocator. 

Mh, I don't see where moving the `Rc` into the `ResourceAllocator` would introduce lifetimes :thinking: (in fact, you cannot have lifetimes here, because you'd get a self-referential struct - if, say, `MmioDeviceManager` gets a reference to the `ResourceAllocator`, then the resource allocator cannot be owned by the `Vmm` struct anymore, as `Vmm` also owns `MmioDeviceManager`. That's why the RC is needed here). You'd simply have multiple `ResourceAllocator`s that point to the same `{Id, Address}Allocator` through the internal `Rc`s. 

> Speaking about "unidiomatism", do you see a practical problem with the way things are? The one I see is that, in theory, someone could create a ResourceAllocator without putting it inside an Rc and I don't know how would that play out.

I don't really see any problems :). A RefCell without an Rc altogether is still a valid usecase. Am happy to keep it the way it is, just had me do a double-take when I started reviewing haha

Reply:
You are right! I hadn't really seen the self-referential issue there. In that case, I would rather go with my approach since it is just one `Rc` shared across all entities that need it, rather than having them share three of those. If you see a functional issue with this, I'm glad to re-consider.